proto/src/xfer/dns_stream.rs

```diff
@@ -11,24 +11,32 @@ use std::io;
 use std::net::SocketAddr;
 
 use futures::stream::{Peekable, Stream};
-use futures::sync::mpsc::{unbounded, UnboundedReceiver};
+use futures::sync::mpsc::{unbounded, UnboundedReceiver, UnboundedSender};
 use futures::{Async, AsyncSink, Future, Poll, Sink};
 
 use error::*;
-use xfer::{SerialMessage, SerialMessageStreamHandle};
+use xfer::{SerialMessage, SerialMessageSender, SerialMessageStreamHandle};
 
 /// TODO: move non-tcp stuff to another type called DNSStream
 /// A Stream used for sending data to and from a remote DNS endpoint (client or server).
 #[must_use = "futures do nothing unless polled"]
-pub struct DnsStream<S> {
+pub struct DnsStream<S, R>
+where
+    S: SerialMessageSender<SerialResponse = R>,
+    R: Future<Item = SerialMessage, Error = io::Error> + Send,
+{
     io_stream: S,
     outbound_messages: Peekable<UnboundedReceiver<SerialMessage>>,
     peer_addr: SocketAddr,
-    to_send: Option<SerialMessage>,
-    is_sending: bool,
+    // TODO: this should just return the future to the requester...
+    requests: Vec<R>,
```

Comment:
I am not happy about this... it's needed to convert the messages sent back into a response stream for the current interfaces. In HTTP2 this isn't necessary, as it's a message/response pattern, where the H2 library is responsible for multiplexing the connections. We need to push the multiplexing logic from DnsFuture down into UdpClientStream and TcpClientStream.